src/local-state/__tests__/LocalState/exports.test.ts

```diff
@@ -0,0 +1,1969 @@
+      { path: ["author", "id"] }
+    )
+  );
+});
+
+test("throws error on @client only queries when the @export directive is used on root field with no associated variable definition", async () => {
+  const document = gql`
+    {
+      field @client @export(as: "someVar")
+    }
+  `;
+
+  const client = new ApolloClient({
+    cache: new InMemoryCache(),
+    link: ApolloLink.empty(),
+  });
+
+  const localState = new LocalState({
+    resolvers: {
+      Query: {
+        field: () => true,
+      },
+    },
+  });
+
+  await expect(
+    localState.getExportedVariables({
+      document,
+      client,
+      context: {},
+      variables: {},
+    })
+  ).rejects.toEqual(
+    new LocalStateError(
+      "`@export` directive on field 'field' does not have an associated variable definition for the 'someVar' variable.",
+      { path: ["field"] }
+    )
+  );
+});
+
+test("throws error on @client only queries when the @export directive is used on nested fields with no associated variable definition", async () => {
+  const document = gql`
+    {
+      car @client {
+        engine {
+          torque @export(as: "torque")
+        }
+      }
+    }
+  `;
+
+  const client = new ApolloClient({
+    cache: new InMemoryCache(),
+    link: ApolloLink.empty(),
+  });
+
+  const localState = new LocalState({
+    resolvers: {
+      Query: {
+        car: () => ({
+          __typename: "Car",
+          engine: {
+            __typename: "Engine",
+            torque: 7200,
+          },
+        }),
+      },
+    },
+  });
+
+  await expect(
+    localState.getExportedVariables({
+      document,
+      client,
+      context: {},
+      variables: {},
+    })
+  ).rejects.toEqual(
+    new LocalStateError(
+      "`@export` directive on field 'torque' does not have an associated variable definition for the 'torque' variable.",
+      { path: ["car", "engine", "torque"] }
+    )
+  );
+});
+
+test("throws error if `@export` variable does not exist in a variable definition when used with server field", async () => {
+  const document = gql`
+    query currentAuthorPostCount {
+      authorId @client @export(as: "authorId")
+      author(id: $authorId) {
+        id
+        name
+      }
+    }
+  `;
+
+  const client = new ApolloClient({
+    cache: new InMemoryCache(),
+    link: ApolloLink.empty(),
+  });
+
+  const testAuthor = {
+    __typename: "Author",
+    id: 100,
+    name: "John Smith",
+  };
+
+  const localState = new LocalState({
+    resolvers: {
+      Query: {
+        author: () => testAuthor,
+        authorId: () => testAuthor.id,
+      },
+    },
+  });
+
+  await expect(
+    localState.getExportedVariables({
+      document,
+      client,
+      context: {},
+      variables: {},
+    })
+  ).rejects.toEqual(
+    new LocalStateError(
+      "`@export` directive on field 'authorId' does not have an associated variable definition for the 'authorId' variable.",
```

Comment:
Maybe we can change the wording here a bit?
```suggestion
      "`@export` directive on field 'authorId' cannot export the $authorId variable as it is missing in the query/mutation/subscription definition.",
```

Reply:
I like this better. Updated in [dd4cd7d](https://github.com/apollographql/apollo-client/pull/12614/commits/dd4cd7d3651f72f8bd2429640879650c6fb0f02b)